File: 30-substring-with-concatenation-of-all-words/substring-with-concatenation-of-all-words.cpp
#include <vector>
#include <string>
#include <unordered_map>

using namespace std;

class Solution {
public:
    vector<int> findSubstring(string s, vector<string>& words) {
        vector<int> result;
        int n = s.length(), m = words.size(), len = words[0].length();
        if (n < m * len) return result;

        unordered_map<string, int> wordCount;
        for (const string& w : words) wordCount[w]++;

        for (int i = 0; i < len; i++) {
            unordered_map<string, int> window;
            int left = i, count = 0;
            for (int j = i; j <= n - len; j += len) {
                string word = s.substr(j, len);
                if (wordCount.count(word)) {
                    window[word]++;
                    count++;
                    while (window[word] > wordCount[word]) {
                        window[s.substr(left, len)]--;
                        count--;
                        left += len;
                    }
                    if (count == m) result.push_back(left);
                } else {
                    window.clear();
                    count = 0;
                    left = j + len;
                }
            }
        }
        return result;
    }
};

File: Substring-with-Concatenation-of-All-Words.java
1import java.util.*;
2
3class Solution {
4    public List<Integer> findSubstring(String s, String[] words) {
5        List<Integer> result = new ArrayList<>();
6        if (s == null || words == null || words.length == 0) return result;
7
8        int wordLen = words[0].length();
9        int wordCount = words.length;
10        int totalLen = wordLen * wordCount;
11        
12        Map<String, Integer> wordMap = new HashMap<>();
13        for (String word : words) {
14            wordMap.put(word, wordMap.getOrDefault(word, 0) + 1);
15        }
16
17        for (int i = 0; i <= s.length() - totalLen; i++) {
18            String sub = s.substring(i, i + totalLen);
19            if (isValid(sub, wordMap, wordLen)) {
20                result.add(i);
21            }
22        }
23
24        return result;
25    }
26
27    private boolean isValid(String sub, Map<String, Integer> wordMap, int wordLen) {
28        Map<String, Integer> seen = new HashMap<>();
29        for (int i = 0; i < sub.length(); i += wordLen) {
30            String word = sub.substring(i, i + wordLen);
31            seen.put(word, seen.getOrDefault(word, 0) + 1);
32            if (seen.get(word) > wordMap.getOrDefault(word, 0)) {
33                return false;
34            }
35        }
36        return true;
37    }
38}
39